### ISSUE TITLE
Question about nLLLoss and CrossEnctropyLoss

### ISSUE THREAD

**Issue by fangfchen** [jxgu1016/MNIST_center_loss.pytorch#12]
Hi, I find the code uses F.log_softmax() and nn.nLLLoss() to count the softmax entropy loss.
And I use nn.CrossEntropyLoss() instead.
But the result is very bad ——test acc is only  93%. 
Could you please tell me what's wrong? Thank you!

**Comment by jxgu1016** [comment]
Please refer to [this issue](https://github.com/jxgu1016/MNIST_center_loss_pytorch/issues/5#issuecomment-385646130).

**Comment by fangfchen** [comment]
Sorry, I feel I can't understand. 
I also use F.log_softmax() and nn.nLLLoss() to train, the acc can get 98%.
But when I use nn.CrossEntropyLoss() instead, the acc is only 93%.
No any other differences. So I'm very confused.

**Comment by fangfchen** [comment]
如果你不介意，我们也可以用中文交流。。。

**Comment by jxgu1016** [comment]
理论上：[CrossEntropyLoss](https://pytorch.org/docs/stable/nn.html#torch.nn.CrossEntropyLoss) = log_softmax + NLLLoss，结果应该是一样的...
不过你改用CrossEntropyLoss的时候，应该把log_softmax去掉了吧？
你能在别的model上复现这个问题吗？

**Comment by fangfchen** [comment]
是的，我把网络的最后改成了：
ip1 = self.preluip1(self.ip1(x))   # the feature dim (to center loss) is 2
ip2 = self.ip2(ip1)
return ip1, ip2
然后损失函数改成了：
softloss = nn.CrossEntropyLoss().to(device)
centerloss = CenterLoss(10, FEAT_DIM).to(device)   # num_class=10, feat dim=2 (feature dim)
使用的时候是：
Ls = softloss(pred, target) 
Lc = centerloss(target, ip1)
loss = Ls + loss_weight * Lc

我用很简单的卷积网络训练mnist，这两种写法是几乎没有区别的。

另外不知道batch_size对centerloss的结果影响大么？（因为center好像是对每个batch算的，如果我的batchsize还没有类别数大的话。。。）

**Comment by fangfchen** [comment]
我刚刚又尝试重新同时跑了这两种写法，发现最后的结果是一样的(acc98.8%)，看来可能是比较受随机因素影响-. -
